README.md
metricsearch
============

standalone graphite metric name search (for usage with custom storage backends)

build instructions:
-------------------
```
git clone git@github.com:viert/metricsearch.git
cd metricsearch
export GOPATH=`pwd`
go get github.com/viert/properties
go get github.com/op/go-logging
go build src/metricsearch.go
```

usage:
------

```
metricsearch 
  -c="/etc/metricsearch.conf": metricsearch config filename
  -reindex="": reindex from plain text metrics file
```

metrics file is a text file with metric names separated by "\n"

metricsearch listens at port 7000 by default and has the following http handlers:

`/add?name=<metricname>` adds metric **metricname** to index, automatically syncing it to disk in background.

`/search?query=<searchquery>` searches for metrics. Metric names are returned line by line, partials (for graphite /metrics/find) are flagged by the following ".". For exapmle:

```
curl "http://localhost:7000/search?query=addressbook.*"
addressbook.host1.
addressbook.host2.
addressbook.total_rps
```
This means host1 and host2 are graphite directories, whereas total_rps is a complete leaf with timeseries.


debian/changelog
metricsearch (0.2-23) precise; urgency=low

  * Pre-insert full tokens validation

 -- Pavel Vorobyov <viert@yandex-team.ru>  Fri, 12 Aug 2016 17:31:43 +0400

metricsearch (0.2-22) precise; urgency=low

  * Pre-insert metric tokens checks
  * Empty token check fixed

 -- Pavel Vorobyov <viert@yandex-team.ru>  Fri, 12 Aug 2016 14:00:52 +0400

metricsearch (0.2-21) precise; urgency=low

  * new validator

 -- Pavel Vorobyov <viert@yandex-team.ru>  Fri, 05 Feb 2016 19:10:41 +0400

metricsearch (0.2-20) precise; urgency=low

  * no stats when self_monitor is off 

 -- Pavel Vorobyov <viert@yandex-team.ru>  Tue, 26 Jan 2016 14:32:53 +0400

metricsearch (0.2-19) precise; urgency=low

  * validation is now optional via main.validate_tokens property (default is
    false)

 -- Pavel Vorobyov <viert@yandex-team.ru>  Wed, 02 Dec 2015 17:28:46 +0400

metricsearch (0.2-18) precise; urgency=low

  * added token validation

 -- Pavel Vorobyov <viert@yandex-team.ru>  Wed, 30 Sep 2015 20:17:46 +0400

metricsearch (0.2-17) precise; urgency=low

  * fixed '?' patter matching

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Thu, 09 Jul 2015 19:35:13 +0300

metricsearch (0.2-16) precise; urgency=low

  * Token length check fixed (?)
  * Removed panic on index file creation errors

 -- Pavel Vorobyov <viert@yandex-team.ru>  Wed, 27 May 2015 16:45:19 +0400

metricsearch (0.2-15) precise; urgency=low

  * Token length check removed temporarily due to unknown strange bug
  * self_monitor_prefix config option added

 -- Pavel Vorobyov <viert@yandex-team.ru>  Wed, 11 Mar 2015 16:46:54 +0400

metricsearch (0.2-14) precise; urgency=low

  * Token length check added

 -- Pavel Vorobyov <viert@yandex-team.ru>  Thu, 22 Jan 2015 16:33:27 +0400

metricsearch (0.2-13) precise; urgency=low

  * Avoid false-positive regexp matches 

 -- Pavel Vorobyov <viert@yandex-team.ru>  Tue, 02 Dec 2014 12:57:26 +0400

metricsearch (0.2-12) precise; urgency=low

  * hostname in self monitoring

 -- Pavel Vorobyov <viert@yandex-team.ru>  Mon, 17 Nov 2014 14:12:47 +0400

metricsearch (0.2-11) precise; urgency=low

  * rps recount fix
  * self monitoring implemented

 -- Pavel Vorobyov <viert@yandex-team.ru>  Mon, 17 Nov 2014 13:39:46 +0400

metricsearch (0.2-10) precise; urgency=low

  * braces match implemented
  * prevent panic on regexp searches

 -- Pavel Vorobyov <viert@yandex-team.ru>  Tue, 11 Nov 2014 20:06:33 +0400

metricsearch (0.2-9) precise; urgency=low

  * stats handler

 -- Pavel Vorobyov <viert@yandex-team.ru>  Thu, 16 Oct 2014 18:00:17 +0400

metricsearch (0.2-8) precise; urgency=low

  * dump handler

 -- Pavel Vorobyov <viert@yandex-team.ru>  Sun, 12 Oct 2014 21:59:23 +0400

metricsearch (0.2-7) precise; urgency=low

  * no_sync

 -- Pavel Vorobyov <viert@yandex-team.ru>  Tue, 07 Oct 2014 19:28:34 +0400

metricsearch (0.2-6) precise; urgency=low

  * logging level fixed?
  * add handler timings 

 -- Pavel Vorobyov <viert@yandex-team.ru>  Tue, 07 Oct 2014 17:36:23 +0400

metricsearch (0.2-5) precise; urgency=low

  * concurrent index writers 

 -- Pavel Vorobyov <viert@yandex-team.ru>  Tue, 07 Oct 2014 13:58:27 +0400

metricsearch (0.2-4) precise; urgency=low

  * mutex locking nodes on update 

 -- Pavel Vorobyov <viert@yandex-team.ru>  Tue, 07 Oct 2014 11:19:59 +0400

metricsearch (0.2-3) precise; urgency=low

  * '?' search 
  * tests

 -- Pavel Vorobyov <viert@yandex-team.ru>  Mon, 06 Oct 2014 16:43:56 +0400

metricsearch (0.2-2) precise; urgency=low

  * disable GC while reindexing and starting up 

 -- Pavel Vorobyov <viert@yandex-team.ru>  Mon, 06 Oct 2014 15:08:15 +0400

metricsearch (0.2-1) precise; urgency=low

  * Minor fixes
  * Go1.3.3 build
  * New release version

 -- Pavel Vorobyov <viert@yandex-team.ru>  Mon, 06 Oct 2014 14:23:06 +0400

metricsearch (0.1-6.1) precise; urgency=low

  * performance rebuild

 -- Pavel Vorobyov <viert@yandex-team.ru>  Mon, 06 Oct 2014 14:15:54 +0400

metricsearch (0.1-6) precise; urgency=low

  * pidfile creation

 -- Pavel Vorobyov <viert@yandex-team.ru>  Mon, 06 Oct 2014 14:00:52 +0400

metricsearch (0.1-5) precise; urgency=low

  * initscript reload enabled 

 -- Pavel Vorobyov <viert@yandex-team.ru>  Mon, 06 Oct 2014 13:45:18 +0400

metricsearch (0.1-4) precise; urgency=low

  * log opening fix
  * background daemon 

 -- Pavel Vorobyov <viert@yandex-team.ru>  Mon, 06 Oct 2014 13:37:06 +0400

metricsearch (0.1-3) precise; urgency=low

  * logrotate
  * initscript

 -- Pavel Vorobyov <viert@yandex-team.ru>  Mon, 06 Oct 2014 13:18:21 +0400

metricsearch (0.1-2) precise; urgency=low

  * stdin indexer
  * index drop before reindex
  * hup signal log release

 -- Pavel Vorobyov <viert@yandex-team.ru>  Mon, 06 Oct 2014 13:11:03 +0400

metricsearch (0.1-1) precise; urgency=low

  * new deafult conffile 

 -- Pavel Vorobyov <viert@yandex-team.ru>  Fri, 03 Oct 2014 18:06:46 +0400

metricsearch (0.1) unstable; urgency=low

  * Initial Release.

 -- Pavel Vorobyov <viert@yandex-team.ru>  Fri, 03 Oct 2014 16:40:15 +0400


src/mstree/mstree_test.go
package mstree

import (
	"bufio"
	"fmt"
	logging "github.com/op/go-logging"
	"os"
	"testing"
	"time"
)

const (
	Data1 = "abook.qa-test1e_yandex_net.some.metric.total"
	Data2 = "abook.qa-test2e_yandex_net.some.metric.total"
	Data3 = "abook.qa-test1d_yandex_net.some.metric.total"
	Data4 = "abook.qa-test2d_yandex_net.some.metric.total"

	DataEmptyToken = "mail.mail_xivahub_var..xivahub.total.1xx"

	LongData = "r9eueKkpYgeezGQsVE5ENyYBpL5XOSgpaskNpNoeVCYfUghzKi89jck8RIcZy3jjOTIQJAAxfpmvFG03Ye2rrTM9c5uup41PCikq8idBObXxBxW07qBtfBP5mxy5MkuhKxCsdnyH06xMn0IF2sILrD9cGu0hFCWs28VwCl4vMifwGd25HIOOgSS3nh8PSsCD34FCgRYcqDnvKVc4s6V0STbTwTIAOHTel3NF56rTETpqAW1Y2XZhP1sbV9VLKMjKq4dfb2Cm7ZZy4JmTGNHtMBEW0M89lQXUnqn4KuLiirENoUaLo33c7L2lh3qWbDXoZQDGIfk7k0cJIL77pP5IKbTCGUEpogSwiuRbwfzhR09F7gZ3x3tDGUliUqV3qWJjtUjw0Qi2w2ixUDSI3OSsacJ90AULlzU8zz8Mbca21odiVuIL2I0uiPxKUOhD3HNdsgdvKugODDCp5acQRNRmoUkp8HkruEVzBCixcyQYdaM7LgbHbJL3i7Jyp3jQ0j8ovhNFrbtoSl074HrPPASrPQPStRWvKbd48dPJwQIfXTydjUmcIwWeEFRXoulA65xGliI1ybqOLXesGOPsaMq5R3Fdn2lFnvmBN1RBZGg4UtABWpRzu.some.valid.tokens"

	InvalidMetric = "'()&%<acx><ScRiPt >prompt(915633)<.if(some){.ops"

	TestExact      = "abook.qa-test1e_yandex_net.some.metric.total"
	TestStarEnd    = "abook.qa-test1e*"
	TestStarBegin  = "abook.*net.some.metric.total"
	TestStarMiddle = "abook.qa-test1*_yandex_net"
	TestStarLonely = "abook.*.some.metric.total"

	TestQuestionEnd     = "abook.qa-test1e_yandex_ne?.some.metric.total"
	TestQuestionBegin   = "abook.?a-test1e_yandex_net.some.metric.total"
	TestQuestionMiddle  = "abook.qa-test1?_yandex_net.some.metric.total"
	TestQuestionFailure = "abook.qa-test?_yandex_net.some.metric.total"

	TestHell    = "abook.q*test?e*.some.*.total"
	TestBraces  = "abook.qa-test[12][ed]*.some.metric.total"
	TestBraces2 = "abook.qa-test[12]e*.some.metric.total"
)

var (
	tree *MSTree
	err  error
)

func prepareTestTree(t testing.TB) {
	if tree != nil {
		return
	}
	dropTestTree()
	tree, err = NewTree("/tmp/test_index", 1000, true)
	if err != nil {
		t.Error(err)
	}
	err = tree.LoadIndex()
	if err != nil {
		t.Error(err)
	}
	tree.Add(Data1)
	tree.Add(Data2)
	tree.Add(Data3)
	tree.Add(Data4)
}

func dropTestTree() {
	os.RemoveAll("/tmp/test_index")
	tree = nil
}

func TestExactMatch(t *testing.T) {
	prepareTestTree(t)
	results := tree.Search(TestExact)
	if len(results) != 1 {
		t.Errorf("Incorrect results length: %d", len(results))
	}
	if results[0] != Data1 {
		t.Errorf("Incorrect result:\n  Got %s\n  Expected %s", results[0], Data1)
	}
}

func TestLongDataDrop(t *testing.T) {
	prepareTestTree(t)
	if tree.TotalMetrics != 4 {
		t.Errorf("Incorrect metrics count on test start:\n  Got %d\n  Expected 4", tree.TotalMetrics)
	}
	tree.Add(LongData)
	if tree.TotalMetrics != 4 {
		t.Errorf("Incorrect metrics count on test end:\n  Got %d\n  Expected 4", tree.TotalMetrics)
	}
}

func TestInvalidDataDrop(t *testing.T) {
	prepareTestTree(t)
	if tree.TotalMetrics != 4 {
		t.Errorf("Incorrect metrics count on test start:\n  Got %d\n  Expected 4", tree.TotalMetrics)
	}
	tree.Add(InvalidMetric)
	if tree.TotalMetrics != 4 {
		t.Errorf("Incorrect metrics count on test end:\n  Got %d\n  Expected 4", tree.TotalMetrics)
	}
}

func TestStarAtTheEnd(t *testing.T) {
	prepareTestTree(t)
	results := tree.Search(TestStarEnd)
	mustMatch := "abook.qa-test1e_yandex_net."
	if len(results) != 1 {
		t.Errorf("Incorrect results length: %d", len(results))
	}
	if results[0] != mustMatch {
		t.Errorf("Incorrect result:\n  Got %s\n  Expected %s", results[0], mustMatch)
	}
}

func TestStarAtTheBegin(t *testing.T) {
	prepareTestTree(t)
	results := tree.Search(TestStarBegin)
	if len(results) != 4 {
		t.Errorf("Incorrect results length: %d", len(results))
	}
}

func TestStarAtTheMiddle(t *testing.T) {
	prepareTestTree(t)
	results := tree.Search(TestStarMiddle)
	if len(results) != 2 {
		t.Errorf("Incorrect results length: %d", len(results))
	}
	for _, metric := range results {
		if metric != "abook.qa-test1d_yandex_net." && metric != "abook.qa-test1e_yandex_net." {
			t.Errorf("Unexpected metric %s", metric)
		}
	}
}

func TestStarLonelyMatch(t *testing.T) {
	prepareTestTree(t)
	results := tree.Search(TestStarLonely)
	if len(results) != 4 {
		t.Errorf("Incorrect results length: %d", len(results))
	}
}

func TestQuestionAtTheEnd(t *testing.T) {
	prepareTestTree(t)
	results := tree.Search(TestQuestionEnd)
	if len(results) != 1 {
		t.Errorf("Incorrect results length: %d", len(results))
	}
	if results[0] != Data1 {
		t.Errorf("Incorrect result:\n  Got %s\n  Expected %s", results[0], Data1)
	}
}

func TestQuestionAtTheBegin(t *testing.T) {
	prepareTestTree(t)
	results := tree.Search(TestQuestionBegin)
	if len(results) != 1 {
		t.Errorf("Incorrect results length: %d", len(results))
	}
	if results[0] != Data1 {
		t.Errorf("Incorrect result:\n  Got %s\n  Expected %s", results[0], Data1)
	}
}

func TestQuestionAtTheMiddle(t *testing.T) {
	prepareTestTree(t)
	results := tree.Search(TestQuestionMiddle)
	if len(results) != 2 {
		t.Errorf("Incorrect results length: %d", len(results))
	}
	for _, metric := range results {
		if metric != Data1 && metric != Data3 {
			t.Errorf("Unexpected metric %s", metric)
		}
	}
}

func TestQuestionNullMatch(t *testing.T) {
	prepareTestTree(t)
	results := tree.Search(TestQuestionFailure)
	if len(results) != 0 {
		for _, metric := range results {
			t.Errorf("Unexpected metric %s", metric)
		}
	}
}

func TestBracesPattern(t *testing.T) {
	prepareTestTree(t)
	results := tree.Search(TestBraces)
	if len(results) != 4 {
		t.Errorf("Incorrect results length: %d", len(results))
	}
}

func TestBracesPattern2(t *testing.T) {
	prepareTestTree(t)
	results := tree.Search(TestBraces2)
	if len(results) != 2 {
		t.Errorf("Incorrect results length: %d", len(results))
	}
}

func TestHellPattern(t *testing.T) {
	prepareTestTree(t)
	results := tree.Search(TestHell)
	if len(results) != 2 {
		t.Errorf("Incorrect results length: %d", len(results))
	}
	for _, metric := range results {
		if metric != Data1 && metric != Data2 {
			t.Errorf("Unexpected metric %s", metric)
		}
	}
}

func TestMetricCount(t *testing.T) {
	prepareTestTree(t)
	if tree.TotalMetrics != 4 {
		t.Errorf("Invalid metrics count after prepare: 4 expected, but %d got", tree.TotalMetrics)
	}
	retries := 10
	for retries > 0 && !tree.Synced() {
		retries--
		time.Sleep(10 * time.Millisecond)
	}
	if !tree.Synced() {
		t.Errorf("Error syncing tree, not synced after 10 retries 10ms each")
	}

	tree, err = NewTree("/tmp/test_index", 1000, true)
	if err != nil {
		t.Error(err)
	}
	tree.LoadIndex()
	if tree.TotalMetrics != 4 {
		t.Errorf("Invalid metrics count after loading index: 4 expected, but %d got", tree.TotalMetrics)
	}
}

func TestEmptyToken(t *testing.T) {
	prepareTestTree(t)
	if tree.TotalMetrics != 4 {
		t.Errorf("Invalid metrics count after prepare: 4 expected, but %d got", tree.TotalMetrics)
	}
	tree.Add(DataEmptyToken)
	if tree.TotalMetrics != 4 {
		t.Errorf("Empty token metric inserted into index!")
	}
}

func BenchmarkTreeAdd(b *testing.B) {
	dropTestTree()
	prepareTestTree(b)
	logging.SetLevel(logging.ERROR, "metricsearch")
	f, err := os.Open("payload.txt")
	if err != nil {
		b.Error("Please provide metric list in payload.txt file in the current directory")
		return
	}
	defer f.Close()

	payload := make([]string, 0, 10000000)
	sc := bufio.NewScanner(f)
	for sc.Scan() {
		payload = append(payload, sc.Text())
	}

	c := 0
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		tree.Add(payload[c])
		c++
		if c == len(payload) {
			fmt.Println("Warning: Payload is too short")
			c = 0
		}
	}
}


src/web/web.go
package web

import (
	"fmt"
	logging "github.com/op/go-logging"
	"io"
	"mstree"
	"net"
	"net/http"
	"os"
	"runtime"
	"strings"
	"sync/atomic"
	"time"
)

type Server struct {
	tree        *mstree.MSTree
	selfMonitor bool
}

type handlerCounters struct {
	add    uint64
	search uint64
	dump   uint64
}

type rpsCounters struct {
	add    float64
	search float64
	dump   float64
}

const (
	monitorHost = "127.0.0.1:42000"
)

var (
	log              *logging.Logger = logging.MustGetLogger("metricsearch")
	totalRequests    handlerCounters
	lastRequests     handlerCounters
	rps              rpsCounters
	monitoringPrefix string
)

func (s *Server) sendMetrics() {
	conn, err := net.Dial("tcp", monitorHost)
	if err != nil {
		return
	}
	defer conn.Close()
	ts := time.Now().Unix()
	sqs, _ := s.tree.SyncQueueSize()
	fmt.Fprintf(conn, "%s.metricsearch.rps.add %.4f %d\n", monitoringPrefix, rps.add, ts)
	fmt.Fprintf(conn, "%s.metricsearch.rps.search %.4f %d\n", monitoringPrefix, rps.search, ts)
	fmt.Fprintf(conn, "%s.metricsearch.rps.dump %.4f %d\n", monitoringPrefix, rps.dump, ts)
	fmt.Fprintf(conn, "%s.metricsearch.reqs.add %.2f %d\n", monitoringPrefix, float32(totalRequests.add), ts)
	fmt.Fprintf(conn, "%s.metricsearch.reqs.search %.2f %d\n", monitoringPrefix, float32(totalRequests.search), ts)
	fmt.Fprintf(conn, "%s.metricsearch.reqs.dump %.2f %d\n", monitoringPrefix, float32(totalRequests.dump), ts)
	fmt.Fprintf(conn, "%s.metricsearch.metrics %.2f %d\n", monitoringPrefix, float64(s.tree.TotalMetrics), ts)
	fmt.Fprintf(conn, "%s.metricsearch.sync_queue %.2f %d\n", monitoringPrefix, float64(sqs), ts)
}

func (s *Server) searchHandler(w http.ResponseWriter, r *http.Request) {
	atomic.AddUint64(&totalRequests.search, 1)
	w.Header().Set("Content-Type", "text/plain")
	r.ParseForm()
	query := r.Form.Get("query")
	tm := time.Now()
	data := s.tree.Search(query)
	dur := time.Now().Sub(tm)
	if dur > time.Millisecond {
		// slower than 1ms
		log.Debug("Searching %s took %s\n", query, dur.String())
	}
	for _, item := range data {
		io.WriteString(w, item+"\n")
	}
}

func (s *Server) addHandler(w http.ResponseWriter, r *http.Request) {
	atomic.AddUint64(&totalRequests.add, 1)
	w.Header().Set("Content-Type", "text/plain")
	r.ParseForm()
	name := r.Form.Get("name")
	if name == "" {
		w.WriteHeader(http.StatusBadRequest)
		io.WriteString(w, "Specify 'name' parameter")
		return
	}
	tm := time.Now()
	s.tree.Add(name)
	dur := time.Now().Sub(tm)
	if dur > time.Millisecond*100 {
		log.Debug("Indexing %s took %s\n", name, dur.String())
	}
	io.WriteString(w, "Ok")
}

func (s *Server) stackHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "text/plain")
	buf := make([]byte, 65536)
	n := runtime.Stack(buf, true)
	w.Write(buf[:n])
}

func (s *Server) dumpHandler(w http.ResponseWriter, r *http.Request) {
	atomic.AddUint64(&totalRequests.dump, 1)
	w.Header().Set("Content-Type", "text/plain")
	s.tree.Root.TraverseDump("", w)
}

func (s *Server) statsHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "text/plain")
	io.WriteString(w, "Total requests (online):\n=============================\n")
	io.WriteString(w, fmt.Sprintf("  add:    %d\n", totalRequests.add))
	io.WriteString(w, fmt.Sprintf("  search: %d\n", totalRequests.search))
	io.WriteString(w, fmt.Sprintf("  dump:   %d\n", totalRequests.dump))
	io.WriteString(w, "\n")
	io.WriteString(w, "RPS (refreshes every minute):\n=============================\n")
	io.WriteString(w, fmt.Sprintf("  add:    %.3f\n", rps.add))
	io.WriteString(w, fmt.Sprintf("  search: %.3f\n", rps.search))
	io.WriteString(w, fmt.Sprintf("  dump:   %.3f\n", rps.dump))
	io.WriteString(w, "\n")
	sqs, _ := s.tree.SyncQueueSize()
	io.WriteString(w, fmt.Sprintf("Total Metrics: %d\n", s.tree.TotalMetrics))
	io.WriteString(w, fmt.Sprintf("Sync Queue Size: %d\n", sqs))
}

func (s *Server) recalcRPS() {
	if s.selfMonitor {
		ticker := time.Tick(time.Minute)
		for _ = range ticker {
			rps.add = float64(totalRequests.add-lastRequests.add) / 60
			rps.dump = float64(totalRequests.dump-lastRequests.dump) / 60
			rps.search = float64(totalRequests.search-lastRequests.search) / 60
			lastRequests = totalRequests
			s.sendMetrics()
		}
	}
}

func NewServer(tree *mstree.MSTree, selfMonitor bool, selfMonitorPrefix string) *Server {
	host, err := os.Hostname()
	if err != nil {
		host = "localhost"
	}
	selfHostname := strings.Replace(host, ".", "_", -1)
	if selfMonitorPrefix != "" {
		monitoringPrefix = fmt.Sprintf("%s.%s", selfMonitorPrefix, selfHostname)
	} else {
		monitoringPrefix = selfHostname
	}
	server := &Server{tree, selfMonitor}
	http.HandleFunc("/search", server.searchHandler)
	http.HandleFunc("/add", server.addHandler)
	http.HandleFunc("/debug/stack", server.stackHandler)
	http.HandleFunc("/dump", server.dumpHandler)
	if selfMonitor {
		http.HandleFunc("/stats", server.statsHandler)
	}
	return server
}

func (s *Server) Start(listenAddr string) {
	log.Notice("Starting background stats job")
	go s.recalcRPS()
	log.Notice("Starting HTTP")
	err := http.ListenAndServe(listenAddr, nil)
	if err != nil {
		log.Error(err.Error())
		panic(err)
	}
}


src/mstree/node.go
package mstree

import (
	"io"
	"regexp"
	"strings"
	"sync"
)

type node struct {
	Children map[string]*node
	Lock     *sync.Mutex
}

func newNode() *node {
	return &node{make(map[string]*node), new(sync.Mutex)}
}

func (n *node) insert(tokens []string, inserted *bool) {
	if len(tokens) == 0 {
		return
	}
	n.Lock.Lock()
	defer n.Lock.Unlock()

	first, tail := tokens[0], tokens[1:]

	child, ok := n.Children[first]
	if !ok {
		*inserted = true
		child = newNode()
		n.Children[first] = child
	}
	child.insert(tail, inserted)
}

func (n *node) TraverseDump(prefix string, writer io.Writer) {
	if len(n.Children) == 0 {
		io.WriteString(writer, prefix+"\n")
	} else {
		for k, node := range n.Children {
			var nPref string
			if prefix == "" {
				nPref = k
			} else {
				nPref = prefix + "." + k
			}
			node.TraverseDump(nPref, writer)
		}
	}
}

func (n *node) search(pattern string) map[string]*node {
	if pattern == "*" {
		return n.Children
	}

	results := make(map[string]*node)

	wcIndex := strings.Index(pattern, "*")
	qIndex := strings.Index(pattern, "?")
	obIndex := strings.Index(pattern, "[")
	cbIndex := strings.Index(pattern, "]")

	if wcIndex == -1 && qIndex == -1 && obIndex == -1 && cbIndex == -1 {
		if node, ok := n.Children[pattern]; ok {
			results[pattern] = node
		}
		return results
	}

	if cbIndex == -1 && obIndex == -1 {
		if qIndex == -1 {
			// Only *
			lwcIndex := strings.LastIndex(pattern, "*")

			if wcIndex != lwcIndex || (wcIndex != 0 && wcIndex != len(pattern)-1) {
				// more than one wildcard or one wildcard in the middle
				rePattern := "^" + strings.Replace(pattern, "*", ".*", -1) + "$"
				re, err := regexp.Compile(rePattern)
				if err != nil {
					return results
				}
				for k, node := range n.Children {
					if re.MatchString(k) {
						results[k] = node
					}
				}
				return results
			}

			if wcIndex == len(pattern)-1 {
				// wildcard at the end
				partial := pattern[:len(pattern)-1]
				for k, node := range n.Children {
					if strings.HasPrefix(k, partial) {
						results[k] = node
					}
				}
			} else {
				// wildcard at the begining
				partial := pattern[1:]
				for k, node := range n.Children {
					if strings.HasSuffix(k, partial) {
						results[k] = node
					}
				}
			}
		} else if wcIndex == -1 {
			// Only ?
			lqIndex := strings.LastIndex(pattern, "?")
			if qIndex != lqIndex || (qIndex != 0 && qIndex != len(pattern)-1) {
				// more than one ? or one ? in the middle
				rePattern := "^" + strings.Replace(pattern, "?", ".", -1) + "$"
				re, err := regexp.Compile(rePattern)
				if err != nil {
					return results
				}
				for k, node := range n.Children {
					if re.MatchString(k) {
						results[k] = node
					}
				}
				return results
			}

			if qIndex == len(pattern)-1 {
				// ? at the end
				partial := pattern[:len(pattern)-1]
				for k, node := range n.Children {
					if k[:len(k)-1] == partial {
						results[k] = node
					}
				}
			} else {
				// ? at the begining
				partial := pattern[1:]
				for k, node := range n.Children {
					if k[1:] == partial {
						results[k] = node
					}
				}
			}

		} else {
			// * and ? presents
			rePattern := "^" + strings.Replace(strings.Replace(pattern, "*", ".*", -1), "?", ".?", -1) + "$"
			re, err := regexp.Compile(rePattern)
			if err != nil {
				return results
			}
			for k, node := range n.Children {
				if re.MatchString(k) {
					results[k] = node
				}
			}
		}
	} else {
		rePattern := "^" + strings.Replace(strings.Replace(pattern, "*", ".*", -1), "?", ".?", -1) + "$"
		re, err := regexp.Compile(rePattern)
		if err != nil {
			return results
		}
		for k, node := range n.Children {
			if re.MatchString(k) {
				results[k] = node
			}
		}
	}

	return results
}


src/metricsearch.go
package main

import (
	"bufio"
	"config"
	"flag"
	"fmt"
	logging "github.com/op/go-logging"
	"mstree"
	"os"
	"os/signal"
	"runtime"
	"runtime/debug"
	"syscall"
	"web"
)

const (
	DEFAULT_CONFIG_FILE = "/etc/metricsearch.conf"
)

var (
	logfileName string
	logfile     *os.File
	log         *logging.Logger = logging.MustGetLogger("metricsearch")
)

func reopenLog() {
	var err error
	if logfile != nil {
		logfile.Close()
	}
	logfile, err = os.OpenFile(logfileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, os.FileMode(0644))
	if err != nil {
		return
	}
	backend := logging.NewLogBackend(logfile, "", 0)
	logging.SetBackend(backend)
}

func hupCatcher() {
	c := make(chan os.Signal)
	signal.Notify(c, syscall.SIGHUP)
	for _ = range c {
		log.Debug("HUP signal catched, reopening logfile %s", logfileName)
		reopenLog()
	}
}

func main() {
	var format string
	var confFile, reindexFile string
	var stdinImport bool
	flag.StringVar(&confFile, "c", DEFAULT_CONFIG_FILE, "metricsearch config filename")
	flag.StringVar(&reindexFile, "reindex", "", "reindex from plain text metrics file")
	flag.BoolVar(&stdinImport, "stdin", false, "reindex from stdin")
	flag.Parse()

	conf := config.Load(confFile)

	switch conf.Log {
	case "syslog":
		backend, err := logging.NewSyslogBackend("metricsearch")
		if err != nil {
			log.Error(err.Error())
			return
		}
		logging.SetBackend(backend)
		format = "%{program} %{level} %{message}"
	case "":
		format = "%{color}%{level} %{color:reset}%{message}"
	default:
		logfileName = conf.Log
		reopenLog()
		go hupCatcher()
		format = "[%{time:2006-01-02 15:04:05}] %{level} %{message}"
	}
	logging.SetFormatter(logging.MustStringFormatter(format))
	logging.SetLevel(conf.LogLevel, "metricsearch")

	tree, err := mstree.NewTree(conf.IndexDirectory, conf.SyncBufferSize, conf.ValidateTokens)
	if err != nil {
		log.Critical("No way to continue, exiting.")
		return
	}

	log.Debug("Configuring runtime: GCPercent(%d), MaxCores(%d), MaxThreads(%d)", conf.GCPercent, conf.MaxCores, conf.MaxThreads)
	runtime.GOMAXPROCS(conf.MaxCores)
	debug.SetGCPercent(conf.GCPercent)
	debug.SetMaxThreads(conf.MaxThreads)

	if stdinImport {
		err := tree.DropIndex()
		if err != nil {
			log.Critical("Error dropping index: %s", err.Error())
			return
		}
		sc := bufio.NewScanner(os.Stdin)
		for sc.Scan() {
			tree.Add(sc.Text())
		}
		log.Notice("Reindexing complete")
		return
	}

	if reindexFile != "" {
		err := tree.DropIndex()
		if err != nil {
			log.Critical("Error dropping index: %s", err.Error())
			return
		}
		err = tree.LoadTxt(reindexFile, -1)
		if err != nil {
			log.Critical("Reindexing error, exiting.")
			return
		} else {
			log.Notice("Reindexing complete")
			return
		}
	} else {
		tree.LoadIndex()
		server := web.NewServer(tree, conf.SelfMonitor, conf.SelfMonitorPrefix)
		addr := fmt.Sprintf("%s:%d", conf.Host, conf.Port)
		server.Start(addr)
	}

}
